pkg/controllermanager/controller/seed/seed_backup_bucket_reconcile.go

```diff
@@ -0,0 +1,186 @@
+// limitations under the License.
+
+package seed
+
+import (
+	"context"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	gardencoreclientset "github.com/gardener/gardener/pkg/client/core/clientset/versioned"
+	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) backupBucketEnqueue(bb *gardencorev1beta1.BackupBucket) {
+	seedName := bb.Spec.SeedName
+	if seedName == nil {
+		return
+	}
+
+	c.seedBackupBucketQueue.Add(*seedName)
+}
+
+func (c *Controller) backupBucketAdd(obj interface{}) {
+	backupBucket, ok := obj.(*gardencorev1beta1.BackupBucket)
+	if !ok {
+		return
+	}
+
+	c.backupBucketEnqueue(backupBucket)
+}
+
+func (c *Controller) backupBucketUpdate(oldObj, newObj interface{}) {
+	oldBackupBucket, ok1 := oldObj.(*gardencorev1beta1.BackupBucket)
+	newBackupBucket, ok2 := newObj.(*gardencorev1beta1.BackupBucket)
+
+	if !ok1 || !ok2 {
+		return
+	}
+
+	if apiequality.Semantic.DeepEqual(oldBackupBucket.Spec, newBackupBucket.Spec) || apiequality.Semantic.DeepEqual(oldBackupBucket.Status, newBackupBucket.Status) {
+		c.backupBucketEnqueue(newBackupBucket)
+	}
+
+	var (
+		oldSeedName = oldBackupBucket.Spec.SeedName
+		newSeedName = newBackupBucket.Spec.SeedName
+	)
+
+	if oldSeedName != nil && (newSeedName == nil || *oldSeedName != *newSeedName) {
+		c.backupBucketEnqueue(oldBackupBucket)
+	}
+}
+
+type backupBucketReconciler struct {
+	clientMap clientmap.ClientMap
+
+	seedLister         gardencorelisters.SeedLister
+	backupBucketLister gardencorelisters.BackupBucketLister
+}
+
+type backupBucketInfo struct {
+	name     string
+	errorMsg string
+}
+
+func (b *backupBucketInfo) String() string {
+	return fmt.Sprintf("Name: %s, Error: %s", b.name, b.errorMsg)
+}
+
+func (b *backupBucketReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	seed, err := b.seedLister.Get(req.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Logger.Debugf("[BACKUPBUCKET SEED RECONCILE] %s - skipping because Seed has been deleted", req.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		logger.Logger.Infof("[BACKUPBUCKET SEED RECONCILE] %s - unable to retrieve seed object from store: %v", req.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	bbs, err := b.backupBucketLister.List(labels.Everything())
```

Comment:
I'm looking forward to the day when we have switched the controllers to controller-runtime so that we can use field selectors here :)